### ISSUE TITLE
TLS handshake error with Registry 2.0.1

### ISSUE THREAD

**Issue by Trefex** [docker/distribution#850]
Dear all,

I have the following problem. When I connect to registry via ATS -> nginx -> registry, I get a TLS handshake error and everything fails. If I connect via nginx -> registry, I get a SSL authority error which is normal since the nginx certs are self-signed. The ATS is mandatory and has a proper certificate as shown below.

Setup:

* apache traffic server (in front of VM, no control, proper certificate)
* nginx 1.9.3 compiled with ldap support (self-signed)
* registry 2.0.1 started as container
* Docker 1.7.1 client and server 

Command to start registry

```bash
docker run -p 127.0.0.1:5000:5000 -v /etc/nginx/certs:/certs -v /mnt/data/docker-data:/data -e REGISTRY_HTTP_TLS_CERTIFICATE=/certs/domain.crt -e REGISTRY_HTTP_TLS_KEY=/certdomain.key -e REGISTRY_STORAGE_FILESYSTEM_ROOTDIRECTORY=/data --name registry registry:2.0.1
```

Attempt to login to new registry

```
~ ❯ docker login docker-r3lab.uni.lu                                                                                                                                                                           06:19:17 ⏎
Username: christophe.trefois
Password:
Email:
Error response from daemon: Unexpected status code [404] : <html>
<head><title>404 Not Found</title></head>
<body bgcolor="white">
<center><h1>404 Not Found</h1></center>
<hr><center>nginx/1.9.3</center>
</body>
</html>
```

Log from boot2docker

```
DEBU[0010] Calling POST /auth
INFO[0010] POST /v1.19/auth
DEBU[0010] pinging registry endpoint https://docker-r3lab.uni.lu/v0/
DEBU[0010] attempting v2 ping for registry endpoint https://docker-r3lab.uni.lu/v2/
DEBU[0010] hostDir: /etc/docker/certs.d/docker-r3lab.uni.lu
DEBU[0010] attempting v1 ping for registry endpoint https://docker-r3lab.uni.lu/v1/
DEBU[0010] hostDir: /etc/docker/certs.d/docker-r3lab.uni.lu
DEBU[0010] Error unmarshalling the _ping RegistryInfo: invalid character '<' looking for beginning of value
DEBU[0010] RegistryInfo.Version: ""
DEBU[0010] Registry standalone header: ''
DEBU[0010] RegistryInfo.Standalone: true
DEBU[0010] attempting v1 login to registry endpoint https://docker-r3lab.uni.lu/v1/
DEBU[0010] hostDir: /etc/docker/certs.d/docker-r3lab.uni.lu
ERRO[0011] Handler for POST /auth returned error: Unexpected status code [404] : <html>
<head><title>404 Not Found</title></head>
<body bgcolor="white">
<center><h1>404 Not Found</h1></center>
<hr><center>nginx/1.9.3</center>
</body>
</html>

ERRO[0011] HTTP Error                                    err=Unexpected status code [404] : <html>
<head><title>404 Not Found</title></head>
<body bgcolor="white">
<center><h1>404 Not Found</h1></center>
<hr><center>nginx/1.9.3</center>
</body>
</html>
 statusCode=404
```

Log from nginx debug error log

https://paste.fedoraproject.org/253931/31035514/

Log from registry container

```
2015/08/11 16:25:14 http: TLS handshake error from 172.17.42.1:57446: tls: first record does not look like a TLS handshake
2015/08/11 16:25:14 http: TLS handshake error from 172.17.42.1:57448: tls: first record does not look like a TLS handshake
2015/08/11 16:25:14 http: TLS handshake error from 172.17.42.1:57450: tls: first record does not look like a TLS handshake
2015/08/11 16:25:14 http: TLS handshake error from 172.17.42.1:57452: tls: first record does not look like a TLS handshake
2015/08/11 16:25:14 http: TLS handshake error from 172.17.42.1:57454: tls: first record does not look like a TLS handshake
2015/08/11 16:25:14 http: TLS handshake error from 172.17.42.1:57456: tls: first record does not look like a TLS handshake
```

Output from a curl 

```
~ ❯ curl -vi https://docker-r3lab.uni.lu/v2/                                                                                                                                                                    06:27:47
* Hostname was NOT found in DNS cache
*   Trying 10.4.0.57...
* Connected to docker-r3lab.uni.lu (10.4.0.57) port 443 (#0)
* TLS 1.2 connection using TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256
* Server certificate: *.uni.lu
* Server certificate: LuxTrust Qualified CA
* Server certificate: LuxTrust root CA
* Server certificate: Baltimore CyberTrust Root
> GET /v2/ HTTP/1.1
> User-Agent: curl/7.37.1
> Host: docker-r3lab.uni.lu
> Accept: */*
>
< HTTP/1.1 502 Invalid HTTP Response
HTTP/1.1 502 Invalid HTTP Response
< Date: Tue, 11 Aug 2015 16:28:23 GMT
Date: Tue, 11 Aug 2015 16:28:23 GMT
< Connection: keep-alive
Connection: keep-alive
* Server ATS/5.2.1 is not blacklisted
< Server: ATS/5.2.1
Server: ATS/5.2.1
< Cache-Control: no-store
Cache-Control: no-store
< Content-Type: text/html
Content-Type: text/html
< Content-Language: en
Content-Language: en
< Content-Length: 256
Content-Length: 256

<
<HTML>
<HEAD>
<TITLE>Web Server Error</TITLE>
</HEAD>

<BODY BGCOLOR="white" FGCOLOR="black">
<H1>Web Server Error</H1>
<HR>

<FONT FACE="Helvetica,Arial"><B>
Description: The host requested  did not return the document correctly.
</B></FONT>
<HR>
</BODY>
* Connection #0 to host docker-r3lab.uni.lu left intact
```

nginx configuration file

https://paste.fedoraproject.org/253935/14393107/

Does anybody have any idea what I can do to debug this or could help me?

Thank you so much,

--
Christophe

**Comment by dmp42** [comment]
Couple of notes:

 * you start your registry on port 5000, but then you try to login without specifying a port. You must `docker login docker-r3lab.uni.lu:5000` if you expect this to work
 * furthermore, both your registry and nginx do use TLS - usually this is not necessary, unless you have a very good reason to do so, and this is definitely the source of many of your problems. I would recommend NOT using TLS in the registry, and sticking with SSL termination on nginx. Or using the latest RC image, and removing nginx entirely.
 * nevertheless, your 502 error is very likely because nginx tries to connect to your registry backend using plain HTTP while your backend registry is indeed requiring TLS. Your nginx configuration is likely at fault here.

**Comment by Trefex** [comment]
Dear @dmp42 

Thanks for the help so far, very insightful. Indeed, point 2 was the big issue here.

Now, however, by enabling basic auth as you mentioned, I receive only 404 errors, and in the nginx debug logs i found the following:

```bash
015/08/11 19:08:25 [info] 14312#14312: *21 no user/password was provided for basic authentication, client: 10.79.2.156, server: docker-r3lab-server.uni.lu, request: "GET /v2/ HTTP/1.1", host: "docker-r3lab-server.uni.lu"
2015/08/11 19:08:25 [debug] 14312#14312: *21 http finalize request: 401, "/v2/?" a:1, c:1
2015/08/11 19:08:25 [debug] 14312#14312: *21 http special response: 401, "/v2/?"
2015/08/11 19:08:25 [debug] 14312#14312: *21 http set discard body
2015/08/11 19:08:25 [debug] 14312#14312: *21 xslt filter header
2015/08/11 19:08:25 [debug] 14312#14312: *21 posix_memalign: 00007FA8B7531F80:4096 @16
2015/08/11 19:08:25 [debug] 14312#14312: *21 HTTP/1.1 401 Unauthorized
Server: nginx/1.9.3
Date: Tue, 11 Aug 2015 19:08:25 GMT
Content-Type: text/html
Content-Length: 194
Connection: keep-alive
WWW-Authenticate: Basic realm="registry.localhost"
Docker-Distribution-Api-Version: registry/2.0.1
```

I did configure the auth basic part like mentioned in the Registry Authentication docs, so not sure what gives.

I would really like to use the nginx container, but we need the ldap module compiled into nginx, so...

Do you have any further ideas / tips for this?

Thanks!

**Comment by Trefex** [comment]
Ok, another update.

So, it didn't work, because strangely enough I need to disable basic auth for GET in order for the ping to work.

Still, now I receive the status code 500 error.

Relevant log from docker daemon 

```bash
DEBU[0556] Calling POST /auth
INFO[0556] POST /v1.19/auth
DEBU[0556] pinging registry endpoint https://docker-r3lab.uni.lu/v0/
DEBU[0556] attempting v2 ping for registry endpoint https://docker-r3lab.uni.lu/v2/
DEBU[0556] hostDir: /etc/docker/certs.d/docker-r3lab.uni.lu
DEBU[0556] attempting v2 login to registry endpoint https://docker-r3lab.uni.lu/v2/
ERRO[0556] Handler for POST /auth returned error: no successful auth challenge for https://docker-r3lab.uni.lu/v2/ - errors: []
ERRO[0556] HTTP Error                                    err=no successful auth challenge for https://docker-r3lab.uni.lu/v2/ - errors: [] statusCode=500
```

**Comment by dmp42** [comment]
@Trefex:

Can you confirm you managed to have a running registry with nginx doing TLS + basic authentication, following the instructions in https://github.com/docker/distribution/blob/master/docs/nginx.md

If that's the case, can you confirm you are able to do the same thing, using your own nginx image (still without LDAP)?

There should be NO modification needed to be added to the example nginx configuration file, specifically none wrt the `/v2` endpoint.

**Comment by yjarosz** [comment]
Hello. I'm also working with @Trefex on this issue.
I think I have found our issue:
what we want is to have a private registry where upload and modification of images are only allowed by our AD. But we want the public to be able to get images from us.

In our nginx configuration file, we have this block:

```bash
limit_except GET HEAD {
      auth_ldap "Forbidden";
      auth_ldap_servers lcsb;
}
```

That prevent the auth challenge to complete:
`docker login docker-r3lab.uni.lu                                                                                                                                    
Username (yohan.jarosz):
FATA[0000] Error response from daemon: no successful auth challenge for https://docker-r3lab.uni.lu/v2/ - errors: []?`

When we comment our `LimitExcept` to apply the auth to the `v2` `location` then it is working as expected.

Is it possible to authenticate every request but not the fetch of images by the public?

**Comment by Trefex** [comment]
@dmp42 it seems that the authentication for v2 is done in a bit strange way. What we mean by that is that docker login simply puts a GET request to `/v2/` which will respond with a 401 and then a challenge to authenticate. However, if we `limit_except` GET then the auth challenge will fail as the GET is allowed but then we are not authorized to do anything else. 

In order to fix this, we now have two locations in nginx, one for `/v2/` and `/v2/` only and one for everything after `/v2/`, so `location /v2/ {}` and `location = /v2/ {}`. Only in the `location /v2/` part we have the `limit_except` and we have full basic auth for `location = /v2/`.

In conclusion, we can now pull without authentication and push only with authentication.

We wished that this could have been handled a bit nicer, eg that docker login is not a GET but rather a POST or something else which would allow for easier filtering. Another option could be that `docker login` would make an API call to an auth endpoint, eg /v2/auth or something.

Thanks for your help!

**Comment by dmp42** [comment]
@Trefex @yjarosz happy you had it work.

No-auth pull with auth-push is definitely a nice use case that others are interested in.

By any chance, could you share your final nginx configuration?

Or... even better... could you submit a separate md doc (like https://github.com/docker/distribution/blob/master/docs/nginx.md ) that explains how to achieve it?

**Comment by stevvooe** [comment]
@Trefex @dmp42 It might be considered a bug that `docker login` uses a GET request to fetch a token. @jlhawn Thoughts?
